### PR TITLE
Impl/rc rpc plugin

### PR DESCRIPTION
**Pull request by BSolonenko**
This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Summary
This PR provides refactoring sdl  and rc rpc plugin's factories 

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by BSolonenko** [comment]
@LuxoftAKutsan @LitvinenkoIra Please review.

File: src/components/application_manager/include/application_manager/command_factory.h [hunk 1]
```diff
@@ -54,6 +58,101 @@ class CommandFactory {
   virtual CommandSharedPtr CreateCommand(
       const commands::MessageSharedPtr& message,
       commands::Command::CommandSource source) = 0;
+  virtual bool IsAbleToProcess(
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/command_factory.h]: @BSolonenko please give doxygen comments to functions

File: src/components/application_manager/include/application_manager/command_factory.h [hunk 2]
```diff
@@ -54,6 +58,101 @@ class CommandFactory {
...
+      const int32_t,
+      const application_manager::commands::Command::CommandSource) const = 0;
+};
+
+class ICommandCreator {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/command_factory.h]: @BSolonenko please give doxygen comments to new classes

File: src/components/application_manager/include/application_manager/command_factory.h [hunk 3]
```diff
@@ -54,6 +58,101 @@ class CommandFactory {
...
+ public:
+  virtual ~ICommandCreator() {}
+  virtual bool isAble() const = 0;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/command_factory.h]: @BSolonenko please give more appropriate name (CanBeCreated for example) of this function and add doxygen comment

File: src/components/application_manager/include/application_manager/command_factory.h [hunk 4]
```diff
@@ -54,6 +58,101 @@ class CommandFactory {
...
+  RPCService& rpc_service_;
+  HMICapabilities& hmi_capabilities_;
+  PolicyHandlerInterface& policy_handler_;
+  bool able_;
+};
+
+struct InvalidCommand {};
+
+template <>
+class CommandCreator<InvalidCommand> : public ICommandCreator {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/command_factory.h]: @BSolonenko Doxygen comments required

File: src/components/application_manager/include/application_manager/command_factory.h [hunk 5]
```diff
@@ -54,6 +58,101 @@ class CommandFactory {
...
+                 HMICapabilities& hmi_capabilities,
+                 PolicyHandlerInterface& policy_handler)
+      : application_manager_(application_manager)
+      , rpc_service_(rpc_service)
+      , hmi_capabilities_(hmi_capabilities)
+      , policy_handler_(policy_handler)
+      , able_(true) {}
+
+ private:
+  bool isAble() const override;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/command_factory.h]: @BSolonenko I think implementation of this function can be moved to ICommandCreator, with variable,

File: src/components/application_manager/include/application_manager/command_factory.h [hunk 6]
```diff
@@ -54,6 +58,101 @@ class CommandFactory {
...
+class ICommandCreator {
+ public:
+  virtual ~ICommandCreator() {}
+  virtual bool isAble() const = 0;
+  virtual CommandSharedPtr create(
+      const commands::MessageSharedPtr& message) const = 0;
+};
+
+template <typename CommandType>
+class CommandCreator : public ICommandCreator {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/command_factory.h]: @BSolonenko 
rename ICommandCreator -> CommandCreator.
CommandCreator -> DefaultCommandCreator

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_on_remote_control_settings_notification.cc [hunk 7]
```diff
@@ -77,40 +79,44 @@ void RCOnRemoteControlSettingsNotification::DisallowRCFunctionality() {
...
-      hmi_apis::Common_RCAccessMode::eType access_mode =
-          hmi_apis::Common_RCAccessMode::INVALID_ENUM;
-      LOG4CXX_DEBUG(logger_, "Allowing RC Functionality");
-      if ((*message_)[app_mngr::strings::msg_params].keyExists(message_params::kAccessMode)) {
-        const std::string access_mode_str =
-            (*message_)[app_mngr::strings::msg_params][message_params::kAccessMode].asString();
+  const bool is_allowed =
+      (*message_)[app_mngr::strings::msg_params][message_params::kAllowed]
+          .asBool();
+  if (is_allowed) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_on_remote_control_settings_notification.cc:95]: @BSolonenko is it qt style fix? Please use script for check style

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc [hunk 8]
```diff
@@ -74,125 +144,125 @@ CommandSharedPtr RCCommandFactory::CreateCommand(
...
-  //          message, app_manager_));
-  //      break;
-  //    }
-  //    default: { break; }
-  //  }
-  //  return command;
+bool RCCommandFactory::IsAbleToProcess(
+    const int32_t function_id,
+    const application_manager::commands::Command::CommandSource message_source)
+    const {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc:203]: @BSolonenko doe not look like standart style. Pleease run checkstyle

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc [hunk 9]
```diff
@@ -74,125 +144,125 @@ CommandSharedPtr RCCommandFactory::CreateCommand(
...
-  //          message, app_manager_));
-  //      break;
-  //    }
-  //    default: { break; }
-  //  }
-  //  return command;
+ICommandCreator& RCCommandFactory::get_mobile_creator_factory(
+    mobile_api::FunctionID::eType id,
+    mobile_api::messageType::eType message_type) const {
+  LOG4CXX_DEBUG(logger_, "CreateMobileCommand function_id: " << id);
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc]: @BSolonenko please log message tyle

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc [hunk 10]
```diff
@@ -74,125 +144,125 @@ CommandSharedPtr RCCommandFactory::CreateCommand(
...
+    default: { return factory.GetCreator<InvalidCommand>(); }
+  }
 }
 
-}  // namespace application_manager
+ICommandCreator& RCCommandFactory::get_hmi_creator_factory(
+    hmi_apis::FunctionID::eType id,
+    hmi_apis::messageType::eType message_type) const {
+  LOG4CXX_DEBUG(logger_, "CreateHMICommand function_id: " << id);
+  CommandCreatorFacotry factory(
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc]: @BSolonenko please use only rc_factory

**Review by LuxoftAKutsan — CHANGES_REQUESTED** [review]

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc [hunk 11]
```diff
@@ -53,7 +53,105 @@
...
+    return false;
+  }
+  CommandSharedPtr create(
+      const commands::MessageSharedPtr& message) const override {
+    UNUSED(message);
+    return CommandSharedPtr();
+  }
+};
+
+struct RCCommandCreatorFacotry {
```
Comment (LitvinenkoIra) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc]: @BSolonenko typo

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/hmi_command_factory.cc [hunk 12]
```diff
@@ -307,3420 +307,1430 @@ HMICommandFactory::HMICommandFactory(
...
-        command.reset(new commands::SDLPolicyUpdate(message,
-                                                    application_manager_,
-                                                    rpc_service_,
-                                                    hmi_capabilities_,
-                                                    policy_handler_));
-      }
-      break;
+      return hmi_apis::messageType::request == message_type
+                 ? factory.GetCreator<commands::SDLPolicyUpdateResponse>()
+                 : factory.GetCreator<commands::SDLPolicyUpdate>();
```
Comment (LitvinenkoIra) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/hmi_command_factory.cc]: @BSolonenko 
```
return hmi_apis::messageType::request == message_type
                ? factory.GetCreator<commands::SDLPolicyUpdate>()
                : factory.GetCreator<commands::SDLPolicyUpdateResponse>();
```

File: src/components/application_manager/include/application_manager/command_factory.h [hunk 13]
```diff
@@ -54,6 +58,98 @@ class CommandFactory {
...
+    return false;
+  }
+  CommandSharedPtr create(
+      const commands::MessageSharedPtr& message) const override {
+    UNUSED(message);
+    return CommandSharedPtr();
+  }
+};
+
+struct CommandCreatorFacotry {
```
Comment (LitvinenkoIra) [inline src/components/application_manager/include/application_manager/command_factory.h]: @BSolonenko CommandCreatorFacotry -> CommandCreatorFactory

**Review by LitvinenkoIra — CHANGES_REQUESTED** [review]

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc [hunk 14]
```diff
@@ -74,125 +174,123 @@ CommandSharedPtr RCCommandFactory::CreateCommand(
...
-  //    }
-  //    default: { break; }
-  //  }
-  //  return command;
+bool RCCommandFactory::IsAbleToProcess(
+    const int32_t function_id,
+    const application_manager::commands::Command::CommandSource message_source)
+    const {
+  using app_mngr::commands::Command;
+  if (Command::SOURCE_HMI == message_source) {
```
Comment (LitvinenkoIra) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc:205]: @BSolonenko if Command::SOURCE_HMI == message_source
we should return get_**hmi**_creator_factory

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc [hunk 15]
```diff
@@ -74,125 +174,123 @@ CommandSharedPtr RCCommandFactory::CreateCommand(
     const app_mngr::commands::MessageSharedPtr& message,
     app_mngr::commands::Command::CommandSource source) {
   if (app_mngr::commands::Command::SOURCE_HMI == source) {
-    return CreateHMICommand(message, source);
+    mobile_apis::messageType::eType message_type =
```
Comment (LitvinenkoIra) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc]: @BSolonenko if if (app_mngr::commands::Command::SOURCE_HMI == source)
we should return get_**hm**i_creator_factory(function_id, message_type).create(message);

**Review by LitvinenkoIra — APPROVED** [review]

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_on_remote_control_settings_notification.cc (hunk 7 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_on_remote_control_settings_notification.cc:95]: @LuxoftAKutsan checked in https://github.com/LuxoftAKutsan/sdl_core/pull/275/commits/5a296e130d1911c156d105e21d0ef00d4c9b97b1

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc (hunk 8 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc:203]: @LuxoftAKutsan checked in https://github.com/LuxoftAKutsan/sdl_core/pull/275/commits/5a296e130d1911c156d105e21d0ef00d4c9b97b1

File: src/components/application_manager/include/application_manager/command_factory.h (hunk 1 above)
Comment (BSolonenko) [reply src/components/application_manager/include/application_manager/command_factory.h]: @LuxoftAKutsan gave in https://github.com/LuxoftAKutsan/sdl_core/pull/275/commits/75224d1cd9b7d4fc21af67229e902ff04fdaedb9

File: src/components/application_manager/include/application_manager/command_factory.h (hunk 2 above)
Comment (BSolonenko) [reply src/components/application_manager/include/application_manager/command_factory.h]: @LuxoftAKutsan gave in https://github.com/LuxoftAKutsan/sdl_core/pull/275/commits/75224d1cd9b7d4fc21af67229e902ff04fdaedb9

File: src/components/application_manager/include/application_manager/command_factory.h (hunk 3 above)
Comment (BSolonenko) [reply src/components/application_manager/include/application_manager/command_factory.h]: @LuxoftAKutsan fixed name and gave comment  in https://github.com/LuxoftAKutsan/sdl_core/pull/275/commits/75224d1cd9b7d4fc21af67229e902ff04fdaedb9

File: src/components/application_manager/include/application_manager/command_factory.h (hunk 4 above)
Comment (BSolonenko) [reply src/components/application_manager/include/application_manager/command_factory.h]: @LuxoftAKutsan  gave in 75224d1

File: src/components/application_manager/include/application_manager/command_factory.h (hunk 6 above)
Comment (BSolonenko) [reply src/components/application_manager/include/application_manager/command_factory.h]: @LuxoftAKutsan  fixed in 75224d1

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc (hunk 9 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc]: @LuxoftAKutsan  fixed in 75224d1

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc (hunk 10 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc]: @LuxoftAKutsan  fixed in 75224d1

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc (hunk 11 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc]: @LuxoftAKutsan  fixed in 75224d1

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/hmi_command_factory.cc (hunk 12 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/hmi_command_factory.cc]: @LuxoftAKutsan  fixed in 75224d1

File: src/components/application_manager/include/application_manager/command_factory.h (hunk 13 above)
Comment (BSolonenko) [reply src/components/application_manager/include/application_manager/command_factory.h]: @LuxoftAKutsan  fixed in 75224d1

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc (hunk 14 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc]: @LuxoftAKutsan  fixed in 75224d1

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc (hunk 15 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc]: @LuxoftAKutsan  fixed in 75224d1

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc (hunk 11 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc]: @LitvinenkoIra   fixed in 75224d1

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/hmi_command_factory.cc (hunk 12 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/hmi_command_factory.cc]: @LitvinenkoIra   fixed in 75224d1

File: src/components/application_manager/include/application_manager/command_factory.h (hunk 13 above)
Comment (BSolonenko) [reply src/components/application_manager/include/application_manager/command_factory.h]: @LitvinenkoIra   fixed in 75224d1

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc (hunk 14 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc:205]: @LitvinenkoIra   fixed in 75224d1

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc (hunk 15 above)
Comment (BSolonenko) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_command_factory.cc]: @LitvinenkoIra   fixed in 75224d1

**Pull request merged**